src/kernel/src/evf/mod.rs

```diff
@@ -0,0 +1,370 @@
+                    let wait_condition_met = match wt.wait_mode {
+                        wm if wm.intersects(EventFlagWaitMode::EVF_WAITMODE_AND) => {
+                            new_val & wt.pattern == wt.pattern
+                        }
+                        wm if wm.intersects(EventFlagWaitMode::EVF_WAITMODE_OR) => {
+                            new_val & wt.pattern != 0
+                        }
+                        _ => todo!("wt.wait_mode does not include neither AND nor OR"),
+                    };
+
+                    if wait_condition_met {
+                        let cleared_val = match wt.wait_mode {
+                            wm if wm.intersects(EventFlagWaitMode::EVF_WAITMODE_CLEAR_ALL) => 0,
+                            wm if wm.intersects(EventFlagWaitMode::EVF_WAITMODE_CLEAR_PAT) => {
+                                new_val & !wt.pattern
+                            }
+                            _ => todo!("wt.wait_mode does not specify CLEAR condition"),
+                        };
+
+                        *pat = cleared_val;
+                        if let Some(arc) = wt.sync.upgrade() {
+                            let (lock, cvar) = &*arc;
+                            let mut matched = lock.lock().unwrap();
+                            *matched = Some(EventFlagCondition::WaitConditionSatisfied(new_val));
+
+                            info!("waking thread {:?} by evf", wt.td.id());
+                            cvar.notify_one();
+                        }
+                    }
+
+                    !wait_condition_met
+                });
+
+                Ok((new_val as usize).into())
+            }
+        }
+    }
+
+    fn sys_evf_clear(self: &Arc<Self>, _: &VThread, i: &SysIn) -> Result<SysOut, SysErr> {
+        let id = i.args[0].into();
+        let pattern: u64 = i.args[1].into();
+
+        info!("sys_evf_clear({}, {:#x})", id, pattern);
+
+        match self.proc.objects_mut().get(id, Some(EventFlag::ENTRY_TYPE)) {
+            None => Err(SysErr::Raw(ESRCH)),
+            Some(entry) => {
+                let flag: &Arc<EventFlag> = &entry
+                    .data()
+                    .clone()
+                    .downcast()
+                    .expect("wrong type of named object");
+
+                let mut pat = flag.pattern.write();
+                let new_val = pattern & *pat;
+                *pat = new_val;
+
+                Ok((new_val as usize).into())
+            }
+        }
+    }
+
+    fn sys_evf_cancel(self: &Arc<Self>, _: &VThread, i: &SysIn) -> Result<SysOut, SysErr> {
+        let id = i.args[0].into();
+        let pattern: u64 = i.args[1].into();
+        let num_waiting_threads: *mut i32 = i.args[2].into();
+
+        info!(
+            "sys_evf_cancel({}, {:#x}, {:#x})",
+            id,
+            pattern,
+            (num_waiting_threads as usize)
+        );
+
+        match self.proc.objects_mut().get(id, Some(EventFlag::ENTRY_TYPE)) {
```

Comment:
This should be 

let entry = self.proc.objects_mut().get(id, Some(EventFlag::ENTRY_TYPE)).ok_or(SysErr::Raw(ESRCH))?;

That gets rid of the match statement